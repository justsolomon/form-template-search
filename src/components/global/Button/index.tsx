import Spinner from '../Spinner';
import styles from './button.module.scss';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** Classname to apply to button to override default styles */
  className?: string;

  /** Boolean showing the loading state of the button */
  isLoading?: boolean;

  /** Boolean showing the disabled state of the button */
  isDisabled?: boolean;

  /** Determines if the button is an icon button */
  iconButton?: boolean;

  /** Variant of the button. Defaults to `solid` */
  variant?: 'solid' | 'ghost' | 'transparent';

  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  children: React.ReactNode;
}

function Button({
  isDisabled,
  isLoading,
  children,
  className = '',
  iconButton,
  variant = 'solid',
  leftIcon,
  rightIcon,
  ...rest
}: ButtonProps) {
  return (
    <button
      className={`${styles['button']} ${className} ${
        styles['button--' + variant]
      } ${iconButton ? styles['button--icon'] : ''} ${
        isLoading ? styles['button--loading'] : ''
      }`}
      disabled={isDisabled || isLoading}
      data-testid="button"
      {...rest}
    >
      {leftIcon ? (
        <span className={styles['button__left-icon']}>{leftIcon}</span>
      ) : null}

      <>
        {isLoading ? (
          <Spinner size="24px" color="#ffffff" data-testid="button__spinner" />
        ) : (
          children
        )}
      </>

      {rightIcon ? (
        <span className={styles['button__right-icon']}>{rightIcon}</span>
      ) : null}
    </button>
  );
}

export default Button;
