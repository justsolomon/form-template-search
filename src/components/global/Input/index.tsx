import styles from './input.module.scss';

interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  leftIcon?: string | React.ReactNode;
  rightIcon?: string | React.ReactNode;
  containerClass?: string;
}

const Input = ({
  leftIcon,
  rightIcon,
  containerClass,
  ...rest
}: InputProps) => {
  return (
    <div className={`${styles['input-group']} ${containerClass}`}>
      {leftIcon ? (
        <div className={styles['input-group__left-element']}>{rightIcon}</div>
      ) : null}

      <input
        className={`${styles['input-group__input']} ${
          leftIcon ? styles['input-group__input--has-left'] : ''
        } ${rightIcon ? styles['input-group__input--has-right'] : ''}`}
        {...rest}
      />

      {rightIcon ? (
        <div className={styles['input-group__right-element']}>{rightIcon}</div>
      ) : null}
    </div>
  );
};

export default Input;
