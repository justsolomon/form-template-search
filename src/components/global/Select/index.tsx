import { ReactComponent as DropdownIcon } from 'assets/vectors/arrow-down.svg';
import styles from './select.module.scss';

interface SelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  label: string;
  options: string[];
}

const Select = ({ label, options, ...rest }: SelectProps) => {
  return (
    <div className={styles['select-group']}>
      <label className={styles['select-group__label']}>{label}</label>

      <select
        className={styles['select-group__select']}
        {...rest}
        data-testid="select"
      >
        {options.map((option, index) => {
          return (
            <option value={option} key={index}>
              {option}
            </option>
          );
        })}
      </select>

      <div className={styles['select-group__dropdown-icon']}>
        <DropdownIcon />
      </div>
    </div>
  );
};

export default Select;
