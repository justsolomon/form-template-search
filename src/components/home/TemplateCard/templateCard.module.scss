@import 'styles/variables';
@import 'styles/mixins';

.template-card {
  @include flex(column);

  justify-content: space-between;
  height: 220px;
  box-shadow: $card-shadow;

  &__details {
    padding: 20px;
    text-transform: capitalize;
  }

  &__header {
    @include truncate-text(2);

    margin-bottom: 18px;
    font-size: $font-size--lg;
    line-height: $font-size--xl;
    color: $text-dark;
  }

  &__description {
    @include truncate-text(3);
  }

  &__footer {
    @include flex-center();

    height: 40px;
    padding: 0 20px;
    background-color: $card-footer-bg;

    &__link {
      text-decoration: none;
      color: $text-green;
    }
  }

  @media #{$media-mobile} {
    height: 200px;

    &__header {
      font-size: $font-size--md;
      line-height: $font-size--lg;
    }

    &__description {
      font-size: $font-size--sm;
      line-height: $font-size--base;
    }
  }
}
