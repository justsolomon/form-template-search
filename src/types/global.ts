export type TemplateCategory = 'Health' | 'E-commerce' | 'Education';
export type TemplateOrder = 'Default' | 'Ascending' | 'Descending';

export interface Template {
  name: string;
  created: Date;
  category: TemplateCategory[];
  description: string;
  link: string;
}
