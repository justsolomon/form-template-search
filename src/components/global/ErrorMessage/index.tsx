import { ReactComponent as ErrorIcon } from 'assets/vectors/error.svg';
import { ReactComponent as RetryIcon } from 'assets/vectors/retry.svg';
import Button from '../Button';
import styles from './errorMessage.module.scss';

interface IErrorMessage {
  error: string;
  retryRequest: () => void;
}

function ErrorMessage({ retryRequest, error }: IErrorMessage) {
  return (
    <div className={styles['error-message']}>
      <ErrorIcon />
      <p
        className={styles['error-message__description']}
        data-testid="error-message__description"
      >
        {error === 'TypeError: Failed to fetch'
          ? 'Looks like you lost your connection. Please check it and try again'
          : error}
      </p>
      <Button
        leftIcon={<RetryIcon />}
        onClick={retryRequest}
        className={styles['error-message__button']}
      >
        Try again
      </Button>
    </div>
  );
}

export default ErrorMessage;
