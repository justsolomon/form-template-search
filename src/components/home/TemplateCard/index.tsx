import { Template } from 'types/global';
import styles from './templateCard.module.scss';

const TemplateCard = ({ name, description, link }: Template) => {
  return (
    <li className={styles['template-card']}>
      <div className={styles['template-card__details']}>
        <h2 className={styles['template-card__header']}>{name}</h2>
        <p className={styles['template-card__description']}>{description}</p>
      </div>
      <div className={styles['template-card__footer']}>
        <a
          href={link}
          className={styles['template-card__footer__link']}
          target="_blank"
          rel="noopenner noreferrer"
        >
          Use Template
        </a>
      </div>
    </li>
  );
};

export default TemplateCard;
