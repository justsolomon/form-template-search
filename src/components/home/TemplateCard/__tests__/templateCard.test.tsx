import { render } from '@testing-library/react';
import { Template } from 'types/global';
import TemplateCard from '..';

const template: Template = {
  category: ['Health', 'E-commerce', 'Education'],
  created: '2022-02-14T11:28:57.155463',
  description: 'Consectetur ullamco cupidatat veniam irure',
  link: 'https://formpl.us/templates',
  name: 'Test template',
};

describe('<TemplateCard />', () => {
  it('should render the TemplateCard component correctly', () => {
    const { asFragment } = render(<TemplateCard {...template} />);

    expect(asFragment).toMatchSnapshot();
  });
});
