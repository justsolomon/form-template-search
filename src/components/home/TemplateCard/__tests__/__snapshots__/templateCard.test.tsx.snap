// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateCard /> should render the TemplateCard component correctly 1`] = `[Function]`;
